README.md
# NAME

stricter - than strict. Fatalize stricter and misc warnings.

# SYNOPSIS

    use stricter;

    my (@a, $x) = (1, 2);
    => (W stricter)(F) Wrong slurpy assignment with @a in LIST, leaving $x uninitialized

    my %h = 0;
    => (W misc)(F) Odd number of elements in hash assignment ERROR

# DESCRIPTION

use stricter adds stricter compile-time checks than strict, enables
the default warnings, enables no multidimensional and fatalizes the
compile-time warnings from this **stricter** and the **misc**
category.

    use stricter;

might be a better replacement for the typical idiom:

    use strict;
    use warnings;

    # plus
    no multidimensional;
    use warnings 'FATAL' => qw(stricter misc);

[common::sense](https://metacpan.org/pod/common::sense) or [strictures](https://metacpan.org/pod/strictures) are similar but have misleading or bad names, and
do not catch wrong slurpy assignments.

stricter adds a new warnings category **stricter**, and throws a warning
on the _"Possibly"_ cases. In the non _"Possibly"_ cases the warnings are **FATAL**.

## Wrong slurpy assignment

When the left-hand side of an list assignment contains an ARRAY or HASH
not as last element.

    my (@a, $x) = (1, 2);
    => (W stricter)(F) Wrong slurpy assignment with @a in LIST, leaving $x uninitialized

    my (%h, $x) = (1, 2);
    => (W stricter)(F) Wrong slurpy assignment with %h in LIST, leaving $x uninitialized

## Possibly missing assignment

When the left-hand side of an list assignment contains not enough elements,
and the right-hand side is a not-empty list it displays a non fatal warning.

    my ($a, $b, $c) = (1, 2);
    => (W stricter) Possibly missing assignment to $c in LIST, leaving $c uninitialized

## Odd number of elements in hash assignment

When the right-hand side of an assignment to a HASH contains an uneven
number of elements, it fatalizes the 'misc' warning.

    my (%h) = (0);
    => (W misc)(F) Odd number of elements in hash assignment

## Use of multidimensional array emulation

no [multidimensional](https://metacpan.org/pod/multidimensional) makes using multidimensional array emulation a
fatal error at compile time. It is mostly confused with `@hash{}` vs `$hash{}`.

    $hash{1, 2};                # (F) Use of multidimensional array emulation
    $hash{join($;, 1, 2)};      # doesn't die

## Relax errors to warnings

The warnings can be **unfatalized** with

    use warnings 'NONFATAL' => 'stricter';

or **hidden** with:

    no warnings qw(stricter misc);

## Legacy note

All those errors are perfectly legal perl syntax, and used quite often
in legacy code. But errors from overseeing such missing
initializations are hard to detect, and should not be allowed in this
stricter mode.

## Better error diagnostics

Unlike most perl 5 errors and warnings `stricter` prints the wrong
variable name, not only the type.

# SEE ALSO

[strict](https://metacpan.org/pod/strict), [warnings](https://metacpan.org/pod/warnings), [perldiag](https://metacpan.org/pod/perldiag), [common::sense](https://metacpan.org/pod/common::sense), [strictures](https://metacpan.org/pod/strictures)

# AUTHOR

Reini Urban, <rurban@cpan.org>

# COPYRIGHT AND LICENSE

Copyright (C) 2015 by cPanel Inc

This library is free software; you can redistribute it and/or modify
it under the terms of the Artistic License 2.0.


Changes
Revision history for the Perl pragma stricter.

0.01  Fri Jul ?? 2015 rurban (not yet released)
	- original version; created by h2xs 1.23 with options
		-X -v 0.01 -f stricter


Makefile.PL
use 5.006; # PL_check added with v5.6
use strict;
use ExtUtils::MakeMaker;
WriteMakefile(
    NAME              => 'stricter',
    VERSION_FROM      => 'stricter.pm',
    PREREQ_PM         => {
        'multidimensional' => 0
    },
    ($] >= 5.005 ?
      (ABSTRACT_FROM  => 'stricter.pm',
       AUTHOR         => 'Reini Urban <rurban@cpan.org>') : ()),
   ($ExtUtils::MakeMaker::VERSION gt '6.46' ?
    ('META_MERGE'  => {
        resources => {
            repository  => 'http://github.com/rurban/stricter',
            license     => 'http://dev.perl.org/licenses/',
        },
     }
    ) : ()),
    SIGN => 1,
);

package MY;
#use Config;

sub depend {
  "
README.md : \$(VERSION_FROM)
	pod2markdown \$(VERSION_FROM) > \$@

release : README.md README dist
	git commit -am\"release \$(VERSION)\"
	echo -n continue?
	read y
	git tag \$(VERSION)
	cpan-upload \$(DISTVNAME).tar\$(SUFFIX)
	git push
	git push --tags

test_cover :: pure_all
	\$(RM_RF) cover_db
	\$(PERLRUNINST) -S cover -test
test_coveralls :: pure_all
	\$(PERLRUNINST) -S cover -test -report coveralls

gcov : \$(BASEEXT).c.gcov \$(BASEEXT).gcov cover_db/\$(BASEEXT)-xs.html

\$(BASEEXT).c.gcov \$(BASEEXT).xs.gcov : \$(BASEEXT).xs
	\$(MAKE) CCFLAGS=\"\$(CCFLAGS) -fprofile-arcs -ftest-coverage\" LDDLFLAGS=\"\$(LDDLFLAGS) -fprofile-arcs -ftest-coverage\"
	gcov \$(BASEEXT).c \$(BASEEXT).xs

cover_db/\$(BASEEXT)-xs.html : \$(BASEEXT).xs.gcov
	PERL5OPT=-MDevel::Cover make test
	-$^X -S gcov2perl \$(BASEEXT).c.gcov \$(BASEEXT).xs.gcov
	$^X -S cover

gprof :
	\$(MAKE) CCFLAGS=\"\$(CCFLAGS) -pg\" LDDLFLAGS=\"\$(LDDLFLAGS) -pg\"
"
}


stricter.pm
package stricter;

use 5.006;
use strict;
our $VERSION = '0.01';
use XSLoader ();
use Carp ();
use warnings (); # load it
use warnings::register;
use multidimensional (); # load it

sub import {
    $^H |= 0x7e2; # use strict
    $^W = 1;      # use warnings 'all', but in C not 100 lines of PP
    warnings->import('FATAL' => qw(stricter misc)); # enable it
    multidimensional->unimport(); # enable it
    $^H{stricter} = 1;
}

sub unimport {
    $^H &= ~(0x7e2); # use strict
    $^W = 0;         # use warnings 'all', but in C not 100 lines of PP
    warnings->import('NONFATAL' => qw(misc)); # disable it
    multidimensional->import(); # disable it
    $^H{stricter} = 0;
}

XSLoader::load;
1;
__END__

=head1 NAME

stricter - than strict. Fatalize stricter and misc warnings.

=head1 SYNOPSIS

  use stricter;

  my (@a, $x) = (1, 2);
  => (W stricter)(F) Wrong slurpy assignment with @a in LIST, leaving $x uninitialized

  my %h = 0;
  => (W misc)(F) Odd number of elements in hash assignment ERROR


=head1 DESCRIPTION

use stricter adds stricter compile-time checks than strict, enables
the default warnings, enables no multidimensional and fatalizes the
compile-time warnings from this B<stricter> and the B<misc>
category.

    use stricter;

might be a better replacement for the typical idiom:

    use strict;
    use warnings;

    # plus
    no multidimensional;
    use warnings 'FATAL' => qw(stricter misc);

L<common::sense> or L<strictures> are similar but have misleading or bad names, and
do not catch wrong slurpy assignments.

stricter adds a new warnings category B<stricter>, and throws a warning
on the I<"Possibly"> cases. In the non I<"Possibly"> cases the warnings are B<FATAL>.

=head2 Wrong slurpy assignment

When the left-hand side of an list assignment contains an ARRAY or HASH
not as last element.

  my (@a, $x) = (1, 2);
  => (W stricter)(F) Wrong slurpy assignment with @a in LIST, leaving $x uninitialized

  my (%h, $x) = (1, 2);
  => (W stricter)(F) Wrong slurpy assignment with %h in LIST, leaving $x uninitialized

=head2 Possibly missing assignment

When the left-hand side of an list assignment contains not enough elements,
and the right-hand side is a not-empty list it displays a non fatal warning.

  my ($a, $b, $c) = (1, 2);
  => (W stricter) Possibly missing assignment to $c in LIST, leaving $c uninitialized

=head2 Odd number of elements in hash assignment

When the right-hand side of an assignment to a HASH contains an uneven
number of elements, it fatalizes the 'misc' warning.

  my (%h) = (0);
  => (W misc)(F) Odd number of elements in hash assignment

=head2 Use of multidimensional array emulation

no L<multidimensional> makes using multidimensional array emulation a
fatal error at compile time. It is mostly confused with C<@hash{}> vs C<$hash{}>.

  $hash{1, 2};                # (F) Use of multidimensional array emulation
  $hash{join($;, 1, 2)};      # doesn't die

=head2  Two-argument open is an error

  open *FH, "<$file";
  => (W misc)(F) Two-argument open is an error

  open my $fh, "|-$file";
  => (W misc)(F) Two-argument open is an error

See L<Perl::Critic::Policy::InputOutput::ProhibitTwoArgOpen>

=head2 Relax errors to warnings

The warnings can be B<unfatalized> with

    use warnings 'NONFATAL' => 'stricter';

or B<hidden> with:

    no warnings qw(stricter misc);

=head2 Legacy note

All those errors are perfectly legal perl syntax, and used quite often
in legacy code. But errors from overseeing such missing
initializations are hard to detect, and should not be allowed in this
stricter mode.

=head2 Better error diagnostics

Unlike most perl 5 errors and warnings C<stricter> prints the wrong
variable name, not only the type.

=head1 SEE ALSO

L<strict>, L<warnings>, L<perldiag>, L<common::sense>, L<strictures>

=head1 AUTHOR

Reini Urban, E<lt>rurban@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by cPanel Inc

This library is free software; you can redistribute it and/or modify
it under the terms of the Artistic License 2.0.

=cut
